tests/subsys/net/lib/lwm2m_fota_utils/src/firmaware.c

```diff
@@ -0,0 +1,808 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/ztest.h>
+#include <zephyr/fff.h>
+#include <zephyr/kernel.h>
+
+#include <net/lwm2m_client_utils.h>
+#include <modem/lte_lc.h>
+#include <modem/modem_info.h>
+#include <zephyr/settings/settings.h>
+#include <modem/modem_info.h>
+#include <modem/nrf_modem_lib.h>
+#include <lwm2m_util.h>
+
+#include "stubs.h"
+
+extern void lwm2m_firmware_emulate_modem_lib_init(int modem_init_ret_val);
+static void tear_down_test(void *fixie);
+static void setup_tear_up(void *fixie);
+static void *init_firmware(void);
+
+ZTEST_SUITE(lwm2m_client_utils_firmware, NULL, init_firmware, setup_tear_up, tear_down_test, NULL);
+
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_ADV_FIRMWARE_UPDATE_OBJ_SUPPORT)
+#define FIRM_UPDATE_OBJECT_ID 33629
+#define INSTANCE_COUNT 2
+static const char app_state[] = "lwm2m:fir/33629/0/3";
+static const char app_result[] = "lwm2m:fir/33629/0/5";
+static const char modem_state[] = "lwm2m:fir/33629/1/3";
+static const char modem_result[] = "lwm2m:fir/33629/1/5";
+static const char modem_type[] = "/33629/1";
+
+static uint8_t app_instance;
+static uint8_t modem_instance = 1;
+#else
+#define FIRM_UPDATE_OBJECT_ID 5
+#define INSTANCE_COUNT 1
+static const char app_state[] = "lwm2m:fir/5/0/3";
+static const char app_result[] = "lwm2m:fir/5/0/5";
+
+static const char modem_state[] = "lwm2m:fir/5/0/3";
+static const char modem_result[] = "lwm2m:fir/5/0/5";
+static const char *modem_type = "/5/0";
+
+static uint8_t app_instance;
+static uint8_t modem_instance;
+#endif
+
+struct firmware_object {
+	uint8_t state;
+	uint8_t result;
+	uint8_t image_type;
+};
+
+static struct firmware_object test_object[2];
+static uint8_t *stream_buf;
+static size_t stream_buf_len;
+static fota_download_callback_t firmware_fota_download_cb;
+
+static struct settings_handler *handler;
+static int fota_download_ret_val;
+static int boot_img_num;
+static bool target_reset_done;
+static bool target_done_result;
+static size_t target_offset;
+static uint8_t test_state_reg;
+
+static int dfu_target_mcuboot_set_buf_stub(uint8_t *buf, size_t len)
+{
+	stream_buf = buf;
+	stream_buf_len = len;
+	return 0;
+}
+
+static const char *lwm2m_adv_path(uint16_t inst, uint16_t res)
+{
+	int ret;
+	static char path[LWM2M_MAX_PATH_STR_LEN];
+
+	ret = snprintk(path, sizeof(path), "%d/%d/%d", FIRM_UPDATE_OBJECT_ID, inst, res);
```

Comment:
Please use fixed-width format specifiers - PRIxx...

Reply:
I'm not a fan using those but if you want

Reply:
I do insist ;)